client/src/partials/budget/analysis/analysis_budget.js

```diff
@@ -9,412 +11,423 @@ angular.module('bhima.controllers')
+  session.mode = 'configuration';
+  session.periods = null;
+  session.selectedPreviousFY = null;
+
+  $scope.months = {
+    0 : 'OPERATING_ACCOUNT.ALL',
+    1 : 'OPERATING_ACCOUNT.JANUARY',
+    2 : 'OPERATING_ACCOUNT.FEBRUARY',
+    3 : 'OPERATING_ACCOUNT.MARCH',
+    4 : 'OPERATING_ACCOUNT.APRIL',
+    5 : 'OPERATING_ACCOUNT.MAY',
+    6 : 'OPERATING_ACCOUNT.JUNE',
+    7 : 'OPERATING_ACCOUNT.JULY',
+    8 : 'OPERATING_ACCOUNT.AUGUST',
+    9 : 'OPERATING_ACCOUNT.SEPTEMBER',
+    10 : 'OPERATING_ACCOUNT.OCTOBER',
+    11 : 'OPERATING_ACCOUNT.NOVEMBER',
+    12 : 'OPERATING_ACCOUNT.DECEMBER'
+  };
+  $scope.total = {};
+
+  dependencies.accounts = {};
+  dependencies.budgets = {
+    query : {
+      tables : {
+        'budget' : { 
+          columns : ['id', 'account_id', 'period_id', 'budget']
           },
+        'period' : {
+          columns : ['fiscal_year_id', 'period_number', 'period_start', 'period_stop', 'locked' ]
+          }
         },
-        orderby : ['fiscal_year.start_year', 'fiscal_year.start_month'],
-      }
-    };
-
-    init();
-
-    function init() {
-      enterprise_id = Number(SessionService.enterprise.id);
-      $scope.enterprise = SessionService.enterprise;
-      dependencies.fiscal_years.query.where = [ 'fiscal_year.enterprise_id=' + enterprise_id ];
-      validate.process(dependencies, ['fiscal_years'])
-      .then(loadFiscalYears);
+      join : [ 'period.id=budget.period_id' ]
     }
+  };
 
-    function addBudgetData() {
-      // Insert the budget numbers into the account data
-      // TODO: The following procedural hacks can be simplified by better SQL queries...
-      // First compute the totals for any accounts with budgets
-      var totalBudget = 0.0;
-      var totalBalance = 0.0;
-      var totals = {};
-      $scope.budgets.data.forEach(function (bud) {
-        if (bud.account_id in totals) {
-          totals[bud.account_id] += bud.budget;
+  dependencies.fiscal_years = {
+    query : {
+      tables : {
+        'fiscal_year' : {
+          columns : ['id', 'fiscal_year_txt', 'start_month', 'start_year', 'previous_fiscal_year']
+        },
+      },
+      orderby : ['fiscal_year.start_year', 'fiscal_year.start_month'],
+    }
+  };
+
+  init();
+
+  function init() {
+    enterprise_id = Number(SessionService.enterprise.id);
+    $scope.enterprise = SessionService.enterprise;
+    dependencies.fiscal_years.query.where = [ 'fiscal_year.enterprise_id=' + enterprise_id ];
+    validate.process(dependencies, ['fiscal_years'])
+    .then(loadFiscalYears);
+  }
+
+  function addBudgetData() {
+    // Insert the budget numbers into the account data
+    // TODO: The following procedural hacks can be simplified by better SQL queries...
+    // First compute the totals for any accounts with budgets
+    var totalBudget = 0.0;
+    var totalBalance = 0.0;
+    var totals = {};
+    $scope.budgets.data.forEach(function (bud) {
+      if (bud.account_id in totals) {
```

Comment:
`for (expression in object) {` can behave strangely.  For example:
```js
var o = {};  assert('toString' in o === true); // this is true!
```
It is better to check if the property exists on the present object by using ` if (!totals[bud.account_id]) { /*.. */ }`.  For example,
```js
$scope.budgets.data.forEach(function (bud) {
  
  // initialize totals to zero
   if (!totals[bud.account_id]) { totals[bud.account_id] = 0 }
   
   // we can now add the totals without doing any other checks!
   totals[bud.account_id] += bud.budget;
   totalBudget += bud.budget;
});
```